frame/balances/src/lib.rs

```diff
@@ -1222,6 +1240,193 @@ impl<T: Config<I>, I: Instance> ReservableCurrency<T::AccountId> for Module<T, I
 	}
 }
 
+impl<T: Config<I>, I: Instance> NamedReservableCurrency<T::AccountId> for Module<T, I>  where
+	T::Balance: MaybeSerializeDeserialize + Debug
+{
+	fn reserved_balance_named(id: &ReserveIdentifier, who: &T::AccountId) -> Self::Balance {
+		let reserves = Self::reserves(who);
+		reserves
+			.binary_search_by_key(id, |data| data.id)
+			.map(|idx| reserves[idx].amount)
+			.unwrap_or_default()
+	}
+
+	/// Move `value` from the free balance from `who` to their reserved balance.
+	///
+	/// Is a no-op if value to be reserved is zero.
+	fn reserve_named(id: &ReserveIdentifier, who: &T::AccountId, value: Self::Balance) -> DispatchResult {
+		<Self as ReservableCurrency<_>>::reserve(who, value)?;
+		Reserves::<T, I>::try_mutate(who, |reserves| -> DispatchResult {
+			match reserves.binary_search_by_key(id, |data| data.id) {
+				Ok(idx) => {
+					// this add can't overflow but just to be defensive.
+					reserves[idx].amount = reserves[idx].amount.saturating_add(value);
+				},
+				Err(idx) => {
+					ensure!((reserves.len() as u32) < T::MaxReserves::get(), Error::<T, I>::TooManyReserves);
+					reserves.insert(idx, ReserveData {
+						id: id.clone(),
+						amount: value
+					});
+				},
+			};
+			Ok(())
+		})
+	}
+
+	/// Unreserve some funds, returning any amount that was unable to be unreserved.
+	///
+	/// Is a no-op if the value to be unreserved is zero.
+	fn unreserve_named(id: &ReserveIdentifier, who: &T::AccountId, value: Self::Balance) -> Self::Balance {
+		if value.is_zero() { return Zero::zero() }
+
+		Reserves::<T, I>::mutate_exists(who, |maybe_reserves| -> Self::Balance {
+			if let Some(reserves) = maybe_reserves.as_mut() {
+				match reserves.binary_search_by_key(id, |data| data.id) {
+					Ok(idx) => {
+						let to_change = cmp::min(reserves[idx].amount, value);
+
+						let remain = <Self as ReservableCurrency<_>>::unreserve(who, to_change);
+
+						// remain should always be zero but just to be defensive here
+						let actual = to_change.saturating_sub(remain);
+
+						// `actual <= to_change` and `to_change <= amount`; qed;
+						reserves[idx].amount -= actual;
+
+						if reserves[idx].amount.is_zero() {
+							if reserves.len() == 1 {
+								// no more named reserves
+								*maybe_reserves = None;
+							} else {
+								// remove this named reserve
+								reserves.remove(idx);
+							}
+						}
+
+						value - actual
+					},
+					Err(_) => {
+						value
+					},
+				}
+			} else {
+				value
+			}
+		})
+	}
+
+	/// Slash from reserved balance, returning the negative imbalance created,
+	/// and any amount that was unable to be slashed.
+	///
+	/// Is a no-op if the value to be slashed is zero.
+	fn slash_reserved_named(
```

Comment:
Could foresee having a `trait NamedReserve` or sth and then dropping the `named` postifx, but then I would be worried that by changing an import (`use traits::ReservableCurrency` -> `use traits::NamedReseve`) the semantics of `T::Currency::slash`. 

But correct me if I am wrong, I think rustc is luckily not smart about this and even if only one of the traits is in scope, it would demand the universal function invocation syntax? 

sth to consider overall.

Reply:
Good idea. I think it is possible to create a `struct NamedReserveAdpator<NamedReservableCurrency, Get<ReserveIdentifier>>` that converts `NamedReservableCurrency` into `ReservableCurrency`

Reply:
Something similar to what we have here https://github.com/open-web3-stack/open-runtime-module-library/blob/2de1e024781a7744a3aeb82229ebcb555163452a/currencies/src/lib.rs#L448